lib/vmtranslator/translator.go

```diff
@@ -0,0 +1,141 @@
+package vmtranslator
+
+import (
+	"log"
+	"strconv"
+	"strings"
+)
+
+type AC_TYPE int
+
+const (
+	AC_OP AC_TYPE = iota
+	AC_RELOP
+	AC_VAR
+	AC_JUMP
+)
+
+type AsmCommand_t struct{}
+
+type AsmCommand interface{}
+
+func TranslateCommand(cmd *Command) (string, error) {
+	log.Fatal("unimplemented")
+	switch cmdType := cmd.cmdType; cmdType {
+	case C_PUSH, C_POP:
+		return pushPopToHack(cmd), nil
+	case C_ARITHMETIC:
+		return arithmeticToHack(cmd), nil
+	default:
+		//ERROR
+		return "", err
+	}
+}
+
+func pushPopToHack(command *Command) (hack string) {
+	if command.cmdType == C_PUSH {
+
+		// Place VM argument data in D register
+		if strings.ToLower(command.arg1) == "constant" {
+			hack = "@" + command.arg2 + "\nD=A\n"
+		} else {
+			hack = vmArgumentAddressToAD(command) + "D=M\n"
+		}
+
+		// Place D in M[SP++]
+		hack += pushFromD()
+
+	} else if command.cmdType == C_POP {
+		// Place VM argument address in M[13] (temp)
+		hack += vmArgumentAddressToAD(command) + tempSaveD("R13")
+
+		// Place M[--SP] in D and store it in M[M[13]] (M[VM argument address])
+		hack += popToD() + tempToA("R13") + "\nM=D"
+
+	} else {
+	} //ERROR
+
+	return hack
+}
+
+func vmArgumentAddressToAD(command *Command) (hack string) {
+	switch command.arg1 {
+	case "local":
+		hack = "@LCL\nAD=M\n"
+	case "argument":
+		hack = "@ARG\nAD=M\n"
+	case "this":
+		hack = "@THIS\nAD=M\n"
+	case "that":
+		hack = "@THAT\nAD=M\n"
+	case "temp":
+		hack = "@5\nD=A\n"
+	case "pointer":
+		hack = "@THIS\nD=A\n"
+	case "static":
+		return "@" + filename + "." + command.arg2 + "\nD=A\n"
+	}
+
+	if arg2 := command.arg2; arg2 == "0" {
+	} else if arg2 == "1" {
+		hack += "AD=A+1\n"
+	} else {
+		hack += "@" + arg2 + "\nAD=D+A\n"
+	}
+
+	return hack
+}
+
+var jmpLabel int = 0
```

Comment:
Maybe declare all constants in one place/block at the top
```suggestion
const jmpLabel int = 0
```

Reply:
i dont think this will work with line 112

Reply:
ah you want it to be static? A unique value every time is is accessed throughout runtime?
statics look awkward...
https://www.lucasepe.it/posts/go-closure-vs-static-variables/